### DISCUSSION TITLE
Using Isar with Freezed package

### DISCUSSION THREAD

**Discussion by dokumanx** [isar/isar#1145] [Show and tell]
Hello everyone,

I've noticed some complaints about using Isar with the Freezed package. Some have suggested separating the data layer for Isar, but I haven't been able to find a good example of how to do that. By separating the layers, we can better manage Isar and Freezed within their respective domains. Here's an example code snippet:

```dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:isar/isar.dart';

part 'timer_exercise.freezed.dart';
part 'timer_exercise.g.dart';

@collection
class TimerExerciseDTO {
  Id? id;
  String? name;
}

extension TimerExerciseExtension on TimerExercise {
  TimerExerciseDTO toDTO() {
    return TimerExerciseDTO()
      ..id = id
      ..name = name;
  }
}

extension TimerExerciseDTOExtension on TimerExerciseDTO {
  TimerExercise toEntity() {
    return TimerExercise(
      id: id!,
      name: name!,
    );
  }

  TimerExercise toEntityInit() {
    return TimerExercise.init(
      id: id!,
      name: name!,
    );
  }
}

@freezed
class TimerExercise with _$TimerExercise {
  const factory TimerExercise({
    required int id,
    required String name,
  }) = _TimerExercise;

  const factory TimerExercise.init({
    required int id,
    @Default('Super Exercise') String name,
  }) = _TimerExerciseInit;

  factory TimerExercise.fromJson(Map<String, dynamic> json) =>
      _$TimerExerciseFromJson(json);

  TimerExercise._();
}

```

`DTO` stands for Data Transfer Object, which is a design pattern used to transfer data between different layers or components of an application. In the context of Isar and Freezed, we can use Isar to retrieve local data and then map it to an Entity class using Freezed.

By mapping the Isar data to an Entity class, we can take advantage of the benefits that Freezed offers, such as immutability and ease of use with named constructors. We can also create extensions to provide additional functionality for the Entity class.

**Comment by frank06** [comment]
I was going to try Isar & Freezed this week!

DTOs are a completely superfluous layer, probably. I've worked in bigger applications (in Java EE back in the day) where DTOs were there just so we could say it was following "design patterns". In reality they only added noise. One good model layer (and a few adapters if needed) is enough.

**Reply by dokumanx** [reply]
I agree with you. The only reason I use it is to workaround the lack of support for collections with Freezed. I separated them so DTOs can stay pure and actual models can be generated by Freezed. But I welcome any better workarounds you can provide. Thank you for your thoughts.

**Reply by flexiglas** [reply]
Hey @frank06, 
gave Isar & Freezed a shot yet? How'd it go? I'm wondering myself, on how to make it work.
